### ISSUE TITLE
Visual Studio 2019 crashes often when editing a blazor project

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/aspnetcore#28301]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/Visual-Studio-2019-crashes-often-when-ed/1267744)._

---
[severity:It's more difficult to complete my work]

Running VS 2019  16.8.2

I have a Blazor solution (.net core 3.1) with 3 projects in it (2x  class, 1 x console app). 

After running the project once or twice, visual studio will crash when editing files. 
These crashes happen about 4-5 times per day. 

When it crashes, typically, I am editing a .cs file (not a .razor file), and while I am typing, the response will slow down. So for example I will type 'protected' but nothing shows on the screen for 4-5 seconds. Then 'protected' appears. Next I will type 'void' and nothing shows up on the screen for 10-15 s. Then 'void' shows up. Then I type somemore, but nothing shows up for about 30s, and then VS crashes.

My system is as as follows: 
- Processor	AMD Ryzen 7 1800X Eight-Core Processor, 3600 Mhz, 8 Core(s), 16 Logical Processor(s)
- BIOS Version/Date	American Megatrends Inc. F23f, 1/05/2018
- SMBIOS Version	3.1
- Embedded Controller Version	255.255
- BIOS Mode	UEFI
- BaseBoard Manufacturer	Gigabyte Technology Co., Ltd.
- BaseBoard Product	AX370-Gaming K7
- BaseBoard Version	Default string
- Platform Role	Desktop
- Secure Boot State	Off
- PCR7 Configuration	Elevation Required to View
- Windows Directory	C:\WINDOWS
- System Directory	C:\WINDOWS\system32
- Boot Device	\Device\HarddiskVolume2
- Locale	United States
- Hardware Abstraction Layer	Version = "10.0.19041.488"
- User Name	AzureAD\JosephTan
- Time Zone	New Zealand Daylight Time
- Installed Physical Memory (RAM)	32.0 GB
- Total Physical Memory	31.9 GB
- Available Physical Memory	23.0 GB
- Total Virtual Memory	36.7 GB
- Available Virtual Memory	22.5 GB
- Page File Space	4.75 GB
- Page File	C:\pagefile.sys

When VS crashes, the following processes will be consuming 2-3 GB each:

- ServiceHub.RoslynCodeAnalysisService.exe
- devenv.exe
- ServiceHub.DataWarehouseHost.exe 

Attached is the crash dump.

---
#### Feedback Bot on 11/24/2020, 07:28 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Dr. Tim dos Santos on 11/27/2020, 00:27 AM: 

<p>I can reproduce this. There was a big performance / memory problem in VS 2019 16.6, that was fixed in one of the latest 16.7 versions; this seems now to have regressed into 16.8.</p><p>Symptom is: devenv.exe and roslyn code analyzer exe taking gigabytes of memory, causing severe slowdowns, and then VS crashes due to OutOfMemoryExceptions.</p>

---
### Original Solutions
(no solutions)

**Comment by NTaylorMullen** [comment]
Also associated: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1250511

**Comment by josephsctan** [comment]
Are there  any workarounds that will prevent this from happening?  VS is crashing around 10 times a day for me, when working on a blazor project. If switching things off will help, I would be keen to try.  
I did try disabling Intellisense, but that doesn't help. 

![image](https://user-images.githubusercontent.com/6090599/101423898-e795df00-395e-11eb-8352-1b606aae446a.png)

**Comment by NTaylorMullen** [comment]
@josephsctan that's awful! I'm sincerely sorry that you're encountering that. Have you found anything that reliably reproduces the issue so we can investigate further? 

If you're willing to be a guinea pig for a bit we've been putting in a lot of effort to building out a new Razor/Blazor editor which is currently available in the form of a preview feature (Tools -> Options -> Environment -> Preview Features):

![image](https://user-images.githubusercontent.com/2008729/87990285-8b5a1d80-ca98-11ea-914f-949fa326d9e3.png)

Give it a shot and let us know if you hit any issues, we're actively trying to improve the Blazor tooling experience!

**Comment by josephsctan** [comment]
> @josephsctan that's awful! I'm sincerely sorry that you're encountering that. Have you found anything that reliably reproduces the issue so we can investigate further?

My current project crashes about 10 times a day. Its pretty (un) reliable ! 
I've submitted crash dumps  when I reported the issue. What else can I send you? 

> If you're willing to be a guinea pig for a bit we've been putting in a lot of effort to building out a new Razor/Blazor editor which is currently available in the form of a preview feature (Tools -> Options -> Environment -> Preview Features):
> Give it a shot and let us know if you hit any issues, we're actively trying to improve the Blazor tooling experience!

Will do.

**Comment by NTaylorMullen** [comment]
> I've submitted crash dumps when I reported the issue. What else can I send you?

Ah, so this issue is associated with your crash. That should be sufficient 😄

**Comment by josephsctan** [comment]
Some notes when using the new razor editor:
- new editor also crashes with out of memory error. after about an hour of using VS (this may be related to the overall crash problem with Blazor, not the new editor)

- intellisense is slow. E.g. when putting my cursor on a <div> tag. it takes 4-5 seconds for the end </div> tag to be highlighted . In the old editor, it is instant.
![image](https://user-images.githubusercontent.com/6090599/101813057-1b504f00-3b81-11eb-9323-642e7bd39162.png)

- style intellisense does not work (in the old one, its pretty snappy)
![image](https://user-images.githubusercontent.com/6090599/101813100-286d3e00-3b81-11eb-89f5-cf0c712ed12b.png)

- some of the syntax highlighting is not working, e.g.  comments, and css  inside <style> tags
- in general, syntax highlighting  is slow

**Comment by VR-Architect** [comment]
I also have this issue about 2-3 times an hour. My project contains Blazor server with 3 class libraries. Each library contains a separate area of the application's gui pages. 

The intellisense seems to not work correctly, VS seems to hang then release. Other times it hangs then restarts. And other times just crashes. It happens so much I had to block the MS crash reporting dll from accessing the internet on my very slow connection. 

Also there is no intellisense on razor pages containing c# code. Code behind seems better when it works. As someone pointed out the Rosyln compiler spikes the CPU.

I thought this was only beta version, so I opened my beta version app in release VS and same issues.

**Comment by NTaylorMullen** [comment]
@josephsctan Out of curiosity what version of VS are you on? We made massive improvements in that area in 16.9 preview builds is why I ask.

@VR-Architect  is this with the official editor or experimental?

**Comment by josephsctan** [comment]
> @josephsctan Out of curiosity what version of VS are you on? We made massive improvements in that area in 16.9 preview builds is why I ask.
16.8.3.. But I will install 16.9 preview 2 and try it now.

**Comment by josephsctan** [comment]
Tried 16.9 preview 2 with the new experimental razor  editor:
* razor syntax highlighting works slowly. 
* css completion does not work at all  (e.g. <div class="....">
* Intellisense performance when editing a .razor file is shockingly bad (5-15 s response time)
See Video : https://drive.google.com/file/d/1g2LjFa2WIfeQBlC91sv2qKyoFRQUfbi2/view?usp=sharing
* using F12 to Go To definition for items in .razor files can sometimes be excruciatingly slow (10-15 s)
* and (most importantly) its still crashing with out of memory errors

When I turned off the experimental razor editor, performance of the above went back to normal (pretty snappy) but VS still crashes with out of memory errors after about an hour's work.

**Comment by NTaylorMullen** [comment]
@josephsctan that's definitely awful! Do you happen to have a solution/project you'd be able to share with us to enable us to more easily investigate the issue? If not we can try and piece things together from the dumps/traces you provided.

**Comment by josephsctan** [comment]
I can't really share the entire code base with you. However, I can describe the solution + projects to see if it helps:
Solution 1 : Class Library (.net standard 2.1) containing  116 .cs files
Solution 2 : Class Library (.net standard 2.1) containing  21 .cs files
Solution 3 : Blazor Server App (.net Core 3.1)  
* syncfusion blazor controls  18.3.0.52
*  285 .cs files
* 119 razor.cs files
* 186 .razor files
* 36 .cshtml files (mostly identity related)
*  www folder: a fairly standard collection of js/css : jquery, bootstrap, font-awesome,  and about 10 js plugins 

Hope that helps. If there are additional diagnostics I can collect for you, please let me know how.

**Comment by Phloog** [comment]
Hey, I did originally also comment on the developercommunity ticket (see above), I just wanted to put a "+1" here as well. This problem (OutOfMemory-Crashes/-Restarts) is really butchering my productivity.

I have a large blazor solution with dozens of areas, libs, tools, etc. It was unusable in 16.6.x, got a lot better in mid 16.7.x, and now in 16.8.x it's really unusable again :-(

My own dumps/diagnostics is attached to the developercommunity ticket @ https://developercommunity.visualstudio.com/content/problem/1289540/vs2019-pro-1683-often-crashes-with-outofmemoryexce.html

Please prioritize this pretty severely, as it seems to affect a lot of people, and makes work very grueling atm.

**Comment by Phloog** [comment]
Interestingly, VS2019 16.8.3 also crashes/restarts when idle (!). If I just let it sit with my solution opened, it eventually crashes/restarts, and repeats this every 30ish minutes. The OOM condition thus seems not to be related to specific user actions in VS (editing, intellisense, ...)

**Comment by NTaylorMullen** [comment]
@Phloog yours actually ended up being different then the original 😄. That being said yours was not something we were aware of although I have [a fix out](https://github.com/dotnet/aspnetcore-tooling/pull/2843). 

@josephsctan's will be fixed in two parts:
1. https://github.com/dotnet/aspnetcore-tooling/pull/2837 (already merged)
2. https://github.com/dotnet/aspnetcore-tooling/pull/2816 (soon to be merged)

**Comment by Phloog** [comment]
@NTaylorMullen That really sounds awersome. Can't wait to test it!

**Comment by Phloog** [comment]
That may be a dumb question, but how can I update the used aspnetcore-tooling in a VS 2019 (preview) installation? Is this even possible? I'd really like to compare the performance without and with this fix.

**Comment by Code-DJ** [comment]
@NTaylorMullen do the fixes apply to VS Code for mac as well?

Since .net 5, I have had random freezes when attaching to the debugger using VS Code.

Steps:
1. run dotnet watch run
2. In VS Code, under Run, use ".NET Core Attach" to attach to the running dll.
3. Sometimes, after making a change, the website doesn't refresh. If you try to hit Ctrl+C to stop "dotnet watch run", it doesn't work. The only option at this point is to open "Activity Monitor" and force quit all "dotnet" processes.

Thanks!

**Comment by NTaylorMullen** [comment]
> That may be a dumb question, but how can I update the used aspnetcore-tooling in a VS 2019 (preview) installation? Is this even possible? I'd really like to compare the performance without and with this fix.

Sooo you could update the Razor portion of your install by building the repo and running the generated VSIX; however, it'd only be 1 part of the solution which wouldn't fix your problems on its own unfortunately. The other part of the problem requires code changes from the dotnet/roslyn repo which has its own requirements on newer versions of VS. 

tl;dr; can't 😢 



> @NTaylorMullen do the fixes apply to VS Code for mac as well?

So these changes would apply to VSCode on all platforms; however, the problem you're running to seems unrelated. I'd recommend following [this guidance](https://github.com/OmniSharp/omnisharp-vscode/blob/master/debugger.md#your-feedback) for getting in touch with the .NET debugger product owner.

**Comment by NTaylorMullen** [comment]
This should be fixed in 16.9-Preview3

**Comment by josephsctan** [comment]
Downloaded 16.9-Preview3 today and can't believe the improvement ! Kudos @NTaylorMullen
